Capri.Services/Proposals/ProposalCreator.cs

```diff
@@ -16,43 +16,91 @@ public class ProposalCreator : IProposalCreator
         }
 
-        public async Task<IServiceResult<Proposal>> Create(ProposalRegistration proposalRegistration)
+        public async Task<IServiceResult<ProposalViewModel>> Create(
+            ProposalRegistration inputData)
         {
-            var currentUser = await _userGetter.GetCurrentUser();
-            var currentPromoter = await _context.Promoters.FirstOrDefaultAsync(p => p.UserId == currentUser.Id);
+            var result = await _userGetter.GetCurrentUser();
+            if(!result.Successful())
+            {
+                var errors = result.GetAggregatedErrors();
+                return ServiceResult<ProposalViewModel>.Error(errors);
+            }
 
-            if (proposalRegistration.Level == StudyLevel.Master)
+            var currentUser = result.Body();
+            var promoter = 
+                await _context
+                .Promoters
+                .Include(p => p.Proposals)
+                .FirstOrDefaultAsync(p => p.UserId == currentUser.Id);
+
+            if(promoter == null)
             {
-                if (_submittedProposalGetter.GetMasterProposalNumber(currentPromoter.Id).Result.Body() >= 
-                    currentPromoter.ExpectedNumberOfMasterProposals)
-                    return ServiceResult<Proposal>.Error("You do not have permissions to create master proposal.");
+                return ServiceResult<ProposalViewModel>.Error("The current user has no associated promoter");
             }
-            else
+
+            if(!HasPermissionToCreateProposal(promoter, inputData.Level))
             {
-                if (_submittedProposalGetter.GetBachelorProposalNumber(currentPromoter.Id).Result.Body() >= 
-                    currentPromoter.ExpectedNumberOfBachelorProposals)
-                    return ServiceResult<Proposal>.Error("You do not have permissions to create bachelor proposal.");
+                return ServiceResult<ProposalViewModel>.Error("You are not allowed to create this type of proposal");
             }
 
-            var proposal = _mapper.Map<Proposal>(proposalRegistration);
-            proposal.Id = Guid.NewGuid();
-            proposal.Promoter = currentPromoter;
-            currentPromoter.Proposals.Add(proposal);
-            _context.Promoters.Update(currentPromoter);
+            var proposal = _mapper.Map<Proposal>(inputData);
+            proposal.Promoter = promoter;
 
+            promoter.Proposals.Add(proposal);
+            _context.Promoters.Update(promoter);
 
             await _context.Proposals.AddAsync(proposal);
             await _context.SaveChangesAsync();
 
-            return ServiceResult<Proposal>.Success(proposal);
+            var proposalViewModel = _mapper.Map<ProposalViewModel>(proposal);
+
+            return ServiceResult<ProposalViewModel>.Success(proposalViewModel);
+        }
+
+        private bool HasPermissionToCreateProposal(Promoter promoter, StudyLevel level)
```

Comment:
Nested if inside switch statement is a very bad code smell. I think it's better to split that if part to new methods.